datastore/providers/pgvector_datastore.py

```diff
@@ -0,0 +1,122 @@
+import os
+from typing import Dict, List, Optional
+
+from pgvector.sqlalchemy import Vector
+from sqlalchemy import Column, Index, String, create_engine
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy.future import select
+from sqlalchemy.orm import sessionmaker
+
+from datastore.datastore import DataStore
+from models.models import (
+    Document,
+    DocumentChunk,
+    DocumentMetadataFilter,
+    QueryResult,
+    QueryWithEmbedding,
+)
+
+Base = declarative_base()
+
+# Read environment variables for pgvector configuration
+PGVECTOR_COLLECTION = os.getenv("PGVECTOR_COLLECTION", "documents")
+PGVECTOR_URL = os.getenv("PGVECTOR_URL")
+
+
+class VectorDocument(DataStore):
+    __tablename__ = PGVECTOR_COLLECTION
+
+    id = Column(String, primary_key=True)
+    document_id = Column(String)
+    text = Column(String)
+    embedding = Column(Vector(1536))
+
+    # Add a Cosine Distance index for faster querying
+    index = Index(
+        "vector_cosine_idx",
+        embedding,
+        postgresql_using="ivfflat",
+        postgresql_ops={"embedding": "vector_cosine_ops", "lists": "100"},
+    )
+
+
+class PgVectorDataStore(DataStore):
+    def __init__(self):
+        # Read the database URL from the PGVECTOR_URL environment variable
+        if not PGVECTOR_URL:
+            raise ValueError("PGVECTOR_URL environment variable is not set")
+
+        self.engine = create_engine(PGVECTOR_URL)
+        Base.metadata.create_all(bind=self.engine)
+        self.session_factory = sessionmaker(bind=self.engine)
+
+    async def _upsert(self, chunks: Dict[str, List[DocumentChunk]]) -> List[str]:
+        with self.session_factory() as session:
+            for document_id, document_chunks in chunks.items():
+                for chunk in document_chunks:
+                    vector_doc = VectorDocument(
+                        id=chunk.id,
+                        document_id=document_id,
+                        text=chunk.text,
+                        embedding=chunk.embedding,
+                    )
+                    session.merge(vector_doc)
+
+            session.commit()
+
+        return list(chunks.keys())
+
+    async def _query(self, queries: List[QueryWithEmbedding]) -> List[QueryResult]:
+        with self.session_factory() as session:
+            results = []
+            for query in queries:
```

Comment:
You need to add support for the optional metadata filter parameter for each query, as mentioned above